File: Example/RoboticusDebugger.cpp
#include "RoboticusDebugger.h"

bool Sensor::isTriggered() const {
  if (operatorStr == ">=")
    return input >= threshold;
  if (operatorStr == "<=")
    return input <= threshold;
  if (operatorStr == ">")
    return input > threshold;
  if (operatorStr == "<")
    return input < threshold;
  if (operatorStr == "==")
    return input == threshold;
  if (operatorStr == "!=")
    return input != threshold;
  return false;
}

RoboticusDebugger::RoboticusDebugger(Print &output)
    : _output(output), _hasData(false) {}

void RoboticusDebugger::begin() {
  _doc.clear();
  _doc["timestamp"] = millis();
  _sensorsArray = _doc["sensors"].to<JsonArray>();
  _vectorsArray = _doc["vectors"].to<JsonArray>();
  _hasData = true;
}

void RoboticusDebugger::addSensor(const String &name, float input,
                                  const String &operatorStr, float threshold,
                                  const String &layer, double x, double y) {
  if (!_hasData)
    begin();

  Sensor sensor{name, input, operatorStr, threshold, layer, x, y};
  add(sensor);
}

void RoboticusDebugger::addVector(const String &name, float rotation,
                                  const String &color, const String &layer,
                                  int x, int y) {
  if (!_hasData)
    begin();

  Vector vector{name, rotation, color, layer, x, y};
  add(vector);
}

void RoboticusDebugger::add(const Sensor &sensor) {
  if (!_hasData)
    return;

  JsonObject obj = _sensorsArray.add<JsonObject>();
  obj["name"] = sensor.name;
  obj["input"] = sensor.input;
  obj["isTriggered"] = sensor.isTriggered();
  obj["threshold"] = sensor.threshold;
  obj["layer"] = sensor.layer;

  JsonObject loc = obj["location"].to<JsonObject>();
  loc["x"] = sensor.x;
  loc["y"] = sensor.y;
}

void RoboticusDebugger::add(const Vector &vector) {
  if (!_hasData)
    return;

  JsonObject obj = _vectorsArray.add<JsonObject>();
  obj["name"] = vector.name;
  obj["rotation"] = vector.rotation;
  obj["color"] = vector.color;
  obj["layer"] = vector.layer;

  JsonObject loc = obj["location"].to<JsonObject>();
  loc["x"] = vector.x;
  loc["y"] = vector.y;
}
void RoboticusDebugger::write() {
  if (!_hasData)
    return;

  _doc.shrinkToFit(); // This is optional according to ArduinoJson docs,
                      // decreases memory usage by a lot.
  serializeJson(_doc, _output);
  _output.println(); // IMPORTANT!! The app parses the json with a new line as a
                     // delimiter between frames, so this is required.
  _hasData = false;
}

bool RoboticusDebugger::isEmpty() const { return !_hasData; }

File: Example/Python/RoboticusDebugger.py
import json
import sys
import time
 
 
_START_TIME = time.monotonic()
 
class Sensor:
    def __init__(self, name: str, input_val: float, operator_str: str,
                 threshold: float, layer: str, x: float, y: float):
        self.name = name
        self.input = float(input_val)
        self.operator_str = operator_str
        self.threshold = float(threshold)
        self.layer = layer
        self.x = x
        self.y = y
 
    def is_triggered(self) -> bool:
        ops = {
            ">=": self.input >= self.threshold,
            "<=": self.input <= self.threshold,
            ">":  self.input >  self.threshold,
            "<":  self.input <  self.threshold,
            "==": self.input == self.threshold,
            "!=": self.input != self.threshold,
        }
        return ops.get(self.operator_str, False)
 
 
class Vector:
    def __init__(self, name: str, rotation: float, color: str,
                 layer: str, x: int, y: int):
        self.name = name
        self.rotation = rotation
        self.color = color
        self.layer = layer
        self.x = x
        self.y = y
 
 
class RoboticusDebugger:
    def __init__(self, output=None):
        self._output = output
        self._sensors: list[dict] = []
        self._vectors: list[dict] = []
        self._has_data = False
 
    def add(self, item):
        if isinstance(item, Sensor):
            self._add_sensor_to_json(item)
        elif isinstance(item, Vector):
            self._add_vector_to_json(item)
        else:
            raise TypeError(f"Expected Sensor or Vector, got {type(item)}")
 
    def add_sensor(self, name, input_val, operator_str, threshold, layer, x, y):
        self.add(Sensor(name, input_val, operator_str, threshold, layer, x, y))
 
    def add_vector(self, name, rotation, color, layer, x, y):
        self.add(Vector(name, rotation, color, layer, x, y))
 
    def write(self):
        if not self._has_data:
            return
 
        frame = {
            "timestamp": self._millis(),
            "sensors":   self._sensors,
            "vectors":   self._vectors,
        }
 
        line = json.dumps(frame, separators=(",", ":")) + "\n"
 
        if self._output is None:
            sys.stdout.write(line)
            sys.stdout.flush()
        else:
            try:
                self._output.write(line.encode())
            except TypeError:
                self._output.write(line)
 
        self._reset()
 
    def is_empty(self) -> bool:
        return not self._has_data
 
    def _reset(self):
        self._sensors = []
        self._vectors = []
        self._has_data = False
 
    def _add_sensor_to_json(self, s: Sensor):
        self._sensors.append({
            "name":        s.name,
            "input":       s.input,
            "isTriggered": s.is_triggered(),
            "threshold":   s.threshold,
            "layer":       s.layer,
            "location":    {"x": s.x, "y": s.y},
        })
        self._has_data = True
 
    def _add_vector_to_json(self, v: Vector):
        self._vectors.append({
            "name":     v.name,
            "rotation": v.rotation,
            "color":    v.color,
            "layer":    v.layer,
            "location": {"x": v.x, "y": v.y},
        })
        self._has_data = True
 
    @staticmethod
    def _millis() -> int:
        return int((time.monotonic() - _START_TIME) * 1000)
 
